core/src/main/java/org/apache/iceberg/encryption/AesGcmInputFile.java

```diff
@@ -20,39 +20,33 @@
 
 import org.apache.iceberg.io.InputFile;
 import org.apache.iceberg.io.SeekableInputStream;
-import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 
 public class AesGcmInputFile implements InputFile {
   private final InputFile sourceFile;
   private final byte[] dataKey;
   private final byte[] fileAADPrefix;
   private long plaintextLength;
 
+  /**
+   * Important: sourceFile.getLength() must return the verified plaintext content length, not the
+   * physical file size after encryption. This protects against tampering with the file size in
+   * untrusted storage systems.
+   */
   public AesGcmInputFile(InputFile sourceFile, byte[] dataKey, byte[] fileAADPrefix) {
     this.sourceFile = sourceFile;
     this.dataKey = dataKey;
     this.fileAADPrefix = fileAADPrefix;
-    this.plaintextLength = -1;
+    this.plaintextLength = sourceFile.getLength();
```

Comment:
A couple of observations

1. AesGcmInputFile objects are produced by the `EncryptionManager.decrypt(EncryptedInputFile encrypted)` function. This function is only given an EncryptedInputFile object, comprised of the source InputFile and its KeyMetadata. So the only way for the decrypt function to get the file length information is to call the source `InputFile.getLength()`. We can't provide the decrypt function with the file length via additional channels.

2. But thats not a problem in the current Iceberg code - because the source InputFile length is fully controlled, and comes from a trusted place (not from the underlying object store). More specifically, the sequence is:

a) **writing**:

_public ManifestFile ManifestWriter.toManifestFile()
    return new GenericManifestFile(file.location(),_   [writer.length()](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestWriter.java#L194),
->
_public long org.apache.iceberg.avro.AvroFileAppender.length()_
       [ return stream.getPos();](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java#L83)

So the length, recorded in the ManifestFile objects for each encrypted manifest, is the plaintext length (and not the actual file length), since the `AesGcmOutputStream.getPos()` returns the plaintext position.

b) **reading**: 

_InputFile ManifestFiles.newInputFile(FileIO io, **ManifestFile manifest**) 
    return InputFile input =_ [io.newInputFile(manifest)](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFiles.java#L349);
->
_public InputFile EncryptingFileIO.newInputFile(ManifestFile manifest) 
    return newDecryptingInputFile(manifest.path(), **manifest.length()**, manifest.keyMetadata());_
->
_public InputFile EncryptingFileIO.newDecryptingInputFile(String path, long plainLength, ByteBuffer buffer) {
    InputFile inputFile = **io.newInputFile(path, plainLength);**
    return em.decrypt(wrap(inputFile, buffer));_

So that the length in the `InputFile sourceFile`, given to the AesGcmInputFile constructor, comes from a safe ManifestFile object (since manifest lists are encrypted/signed), and not from an unsafe storage.

What do you think?

Reply:
I think the problem isn't just this path. It is with other valid uses of the API.

For instance, if you write data into an encrypted Avro file and go to read it, it is valid to open the Avro file without calling `newInputFile(DataFile)`. You could use `newInputFile(String)` and that `InputFile` would return the file length from the object store, which is the encrypted length. If the decryption stream assumes here that the incoming `InputFile` reports the plaintext length, there's no way to know or guarantee it.

I think this needs to assume that the incoming file uses the encrypted length.

We still need to choose whether to store the plaintext length or the encrypted length in table metadata. From your comment here, it looks like it would be easier to store the plaintext length because that is what gets returned by the writer. I think that makes sense, but we will need to be specific in `EncryptingFileIO` what is passed through and where it needs to be converted.

Reply:
Another problem I just ran into is that if we store the plaintext length, then `EncryptingFileIO` doesn't know what length to pass into `FileIO.newInputFile(String, long)` for the encrypted file length. Encrypted length depends on the encryption scheme and we don't know the details of the scheme used by some `EncryptionManager`.

That makes me think that we should store the encrypted length in Iceberg metadata. That is also intuitive because the metadata length always matches the object storage length.

We'll have to find out how to get the encrypted length in the data and manifest write paths.

Reply:
Probably the `org.apache.iceberg.io.PositionOutputStream` - then we can have a default impl that returns `getPos()` - so we don't have to modify all Iceberg stream classes. TBD.